src/Symfony/Component/Encryption/EncryptionInterface.php

```diff
@@ -0,0 +1,137 @@
+     *     $message = $encryption->decrypt($ciphertext, $key);
+     * </code>
+     *
+     * @param string       $message plain text version of the message
+     * @param KeyInterface $myKey   a key that holds a string secret
+     *
+     * @return string the output
+     *
+     * @throws EncryptionException
+     */
+    public function encrypt(string $message, KeyInterface $myKey): string;
+
+    /**
+     * Get an encrypted version of the message that only the recipient can read.
+     *
+     * Asymmetric encryption uses a "key pair" ie a public key and a private key.
+     * It is safe to share your public key, but the private key should always be
+     * kept a secret.
+     *
+     * When Alice and Bob wants to communicate they share their public keys with
+     * each other. Alice will encrypt a message with bobs public key. When Bob
+     * receive the message, he will decrypt it with his private key.
+     *
+     *
+     * <code>
+     *     $bobKey = $encryption->generateKey();
+     *     $bobPublicKey = serialize($bobKey->createPublicKey());
+     *
+     *     // Bob sends the public key to Alice
+     *
+     *     $key = unserialize($bobPublicKey);
+     *     $ciphertext = $encryption->encryptFor('input', $key);
+     *
+     *     // Alice sends ciphertext to bob
+     *
+     *     $message = $encryption->decrypt($ciphertext, $bobKey);
+     * </code>
+     *
+     * @param string       $message      plain text version of the message
+     * @param KeyInterface $recipientKey A key with a public key of the recipient
+     *
+     * @return string the output
+     *
+     * @throws EncryptionException
+     */
+    public function encryptFor(string $message, KeyInterface $recipientKey): string;
+
+    /**
+     * Get an encrypted version of the message that only the recipient can read.
+     * The recipient can also verify who sent the message.
+     *
+     * Asymmetric encryption uses a "key pair" ie a public key and a private key.
+     * It is safe to share your public key, but the private key should always be
+     * kept a secret.
+     *
+     * When Alice and Bob wants to communicate they share their public keys with
+     * each other. Alice will encrypt a message with keypair [ alice_private, bob_public ].
+     * When Bob receive the message, he will decrypt it with keypair [ bob_private, alice_public ].
+     *
+     * <code>
+     *     $aliceKey = $encryption->generateKey();
+     *     $bobKey = $encryption->generateKey();
+     *     $keypair = $aliceKey->createKeypair($bobKey);
+     *
+     *     $ciphertext = $encryption->encryptForAndSign('input', $keypair);
+     *     $message = $encryption->decrypt($ciphertext, $bobKey->createKeypair($aliceKey));
+     * </code>
+     *
+     * @param string       $message plain text version of the message
+     * @param KeyInterface $keypair A key with a public key of the recipient and a private key of the sender
```

Comment:
Using `keypair` as name looks wrong to me. In all crypto articles I read, `keypair` refers to the pair of a private key and its associated public key.
Here, you use this for 2 unrelated keys (and you even mix them in the same object)

Reply:
Hmm. Im not sure if you are correct or not. 

What I want to refer to is a "pair of keys". It does not have to be a private keys corresponding public key. But it must be a private/public combination. 

Maybe find another name to make this more clear?

Reply:
or maybe we should pass the recipient key and the sender key as separate arguments

Reply:
That would skip the "Key pair" issue.. Correct. Hm.. Yes that might be simpler

Good idea!

Reply:
But.. That also means that we need two keys for the `decrypt()`. 

Should I update it to : 

```php
EncryptionInterface::decrypt(string $message, KeyInterface $key, KeyInterface $senderPublicKey = null);
```
??

Reply:
I'm wondering whether it is a good idea to have a single `decrypt` method covering the 3 encryption APIs. Implementations have to guess which of the 3 APIs was used to encrypt, because the decryption logic is different in each case.

Reply:
Both yes and no. 

The implementation will know what algorithm that has been used. Ie, it got a "routing key", so it does not have to guess.